extension/healthcheckv2extension/internal/status/aggregator.go

```diff
@@ -0,0 +1,225 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package status // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckv2extension/internal/status"
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+// Extensions are treated as a pseudo pipeline and extsID is used as a map key
+var (
+	extsID    = component.MustNewID("extensions")
+	extsIDMap = map[component.ID]struct{}{extsID: {}}
+)
+
+// Note: this interface had to be introduced because we need to be able to rewrite the
+// timestamps of some events during aggregation. The implementation in core doesn't currently
+// allow this, but this interface provides a workaround.
+type Event interface {
+	Status() component.Status
+	Err() error
+	Timestamp() time.Time
+}
+
+// Scope refers to a part of an AggregateStatus. The zero-value, aka ScopeAll,
+// refers to the entire AggregateStatus. ScopeExtensions refers to the extensions
+// subtree, and any other value refers to a pipeline subtree.
+type Scope string
+
+const (
+	ScopeAll        Scope  = ""
+	ScopeExtensions Scope  = "extensions"
+	pipelinePrefix  string = "pipeline:"
+)
+
+func (s Scope) toKey() string {
+	if s == ScopeAll || s == ScopeExtensions {
+		return string(s)
+	}
+	return pipelinePrefix + string(s)
+}
+
+type Verbosity bool
+
+const (
+	Verbose Verbosity = true
+	Concise           = false
+)
+
+// AggregateStatus contains a map of child AggregateStatuses and an embedded Event.
+// It can be used to represent a single, top-level status when the ComponentStatusMap
+// is empty, or a nested structure when map is non-empty.
+type AggregateStatus struct {
+	Event
+
+	ComponentStatusMap map[string]*AggregateStatus
+}
+
+func (a *AggregateStatus) clone(verbosity Verbosity) *AggregateStatus {
+	st := &AggregateStatus{
+		Event: a.Event,
+	}
+
+	if verbosity == Verbose && len(a.ComponentStatusMap) > 0 {
+		st.ComponentStatusMap = make(map[string]*AggregateStatus, len(a.ComponentStatusMap))
+		for k, cs := range a.ComponentStatusMap {
+			st.ComponentStatusMap[k] = cs.clone(verbosity)
+		}
+	}
+
+	return st
+}
+
+type subscription struct {
+	statusCh  chan *AggregateStatus
+	verbosity Verbosity
+}
+
+// Aggregator records individual status events for components and aggregates statuses for the
+// pipelines they belong to and the collector overall.
+type Aggregator struct {
+	mu              sync.RWMutex
```

Comment:
To confirm: this is intended to guard both `aggregateStatus` and `subscriptions`, so that no concurrent updates to subscriptions and aggregateStatus can happen, right?

Reply:
Correct